File: mobile/modules/acs-meeting/android/src/main/java/com/mentra/acsmeeting/source/GlassesMediaSource.kt
package com.mentra.acsmeeting.source

enum class SourceKind { WHEP, DIRECT }

enum class SourceState { IDLE, CONNECTING, LIVE, FAILED }

data class SourceConfig(
  val url: String,
  val kind: SourceKind = SourceKind.WHEP,
)

/** ACS-negotiated output size. One object so a frame cannot observe a torn width/height. */
data class TargetSize(val width: Int, val height: Int)

/**
 * Whether a subscriber already on [currentUrl] can serve [next] as-is instead of
 * being rebuilt.
 *
 * Reuse is the whole point: a rebuild mints a new peer and restarts the wait for
 * the first frame, which is the outage we are trying to shorten. `CONNECTING`
 * counts as reusable, and since `LIVE` means "a frame reached the sink" that now
 * spans the answer → first frame window too. It cannot strand a caller, because
 * every `CONNECTING` is bounded: the offer post fails, or the answer's
 * first-frame deadline expires into `FAILED` and this returns false.
 */
fun canReuseSource(currentUrl: String?, state: SourceState, next: SourceConfig): Boolean =
  currentUrl == next.url && next.kind == SourceKind.WHEP && state != SourceState.FAILED

/**
 * Observes source health transitions. `FAILED` is the one that matters: ICE
 * dropped, the WHEP endpoint went away, or an answered subscription never
 * delivered a frame — and nothing downstream will notice on its own, because
 * ACS keeps the call up with a frozen last frame.
 */
fun interface SourceStateListener {
  fun onSourceState(state: SourceState, reason: String?)
}

/**
 * Transports whatever the glasses already captured. Capture policy lives
 * outside this interface so a Cloudflare hop can be replaced without touching ACS.
 */
interface GlassesMediaSource {
  fun start(config: SourceConfig)
  fun restart(config: SourceConfig)
  /** Rebuild the transport for the current config even if it looks healthy. */
  fun forceRestart() {}
  fun stop()
  val state: SourceState
  fun setPcmDeliveryEnabled(enabled: Boolean)
  fun setTargetSize(size: TargetSize?) {}
  fun setStateListener(listener: SourceStateListener?) {}
}

fun interface GlassesMediaSourceFactory {
  fun create(video: VideoFrameListener, pcm: PcmListener): GlassesMediaSource
}

class GlassesMediaController(
  private val factory: GlassesMediaSourceFactory,
) {
  private var source: GlassesMediaSource? = null
  private var stateListener: SourceStateListener? = null

  val state: SourceState
    get() = source?.state ?: SourceState.IDLE

  fun attach(video: VideoFrameListener, pcm: PcmListener, config: SourceConfig) {
    source?.stop()
    source = factory.create(video, pcm).also {
      it.setStateListener(stateListener)
      it.start(config)
    }
  }

  fun restart(config: SourceConfig) {
    source?.restart(config)
  }

  fun forceRestart() {
    source?.forceRestart()
  }

  fun setStateListener(listener: SourceStateListener?) {
    stateListener = listener
    source?.setStateListener(listener)
  }

  fun stop() {
    source?.stop()
    source = null
  }

  fun setPcmDeliveryEnabled(enabled: Boolean) {
    source?.setPcmDeliveryEnabled(enabled)
  }

  fun setTargetSize(size: TargetSize?) {
    source?.setTargetSize(size)
  }
}

File: mobile/modules/acs-meeting/ios/PolicyKit/Sources/AcsAudioPolicy/GlassesMediaSource.swift
import Foundation

public enum SourceKind: String, Sendable {
  case whep
  case direct
}

public enum SourceState: String, Sendable {
  case idle
  case connecting
  case live
  case failed
}

public struct SourceConfig: Equatable, Sendable {
  public var url: String
  public var kind: SourceKind

  public init(url: String, kind: SourceKind = .whep) {
    self.url = url
    self.kind = kind
  }
}

public protocol GlassesMediaSource: AnyObject {
  var state: SourceState { get }
  func start(config: SourceConfig)
  func restart(config: SourceConfig)
  func stop()
  func setPcmDeliveryEnabled(_ enabled: Bool)
}

public final class GlassesMediaController {
  private let factory: () -> GlassesMediaSource
  private var source: GlassesMediaSource?

  public init(factory: @escaping () -> GlassesMediaSource) {
    self.factory = factory
  }

  public var state: SourceState { source?.state ?? .idle }

  public func attach(config: SourceConfig) {
    source?.stop()
    let next = factory()
    source = next
    next.start(config: config)
  }

  public func restart(config: SourceConfig) {
    source?.restart(config: config)
  }

  public func stop() {
    source?.stop()
    source = nil
  }

  public func setPcmDeliveryEnabled(_ enabled: Bool) {
    source?.setPcmDeliveryEnabled(enabled)
  }
}

public enum CapturePolicy {
  public static func captureGlassesMic(_ source: AudioSourceKind) -> Bool {
    source == .glasses
  }
}
